internal/environment/environment.go

```diff
@@ -0,0 +1,161 @@
+// The environment package contains probing functionality to
+// detect information on the Kubernetes cluster environment.
+package environment
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/discovery"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	manifestsv1alpha1 "package-operator.run/apis/manifests/v1alpha1"
+)
+
+var _ manager.Runnable = (*Manager)(nil)
+
+const (
+	environmentProbeInterval    = 1 * time.Minute
+	openShiftClusterVersionName = "version"
+	openShiftProxyName          = "cluster"
+)
+
+type Sink interface {
+	InjectEnvironment(env manifestsv1alpha1.PackageEnvironment)
+}
+
+type Manager struct {
+	client          client.Client
+	discoveryClient discovery.DiscoveryInterface
+	sinks           []Sink
+}
+
+func NewManager(
+	client client.Client,
+	discoveryClient discovery.DiscoveryInterface,
+	sinks []Sink,
+) *Manager {
+	return &Manager{
+		client:          client,
+		discoveryClient: discoveryClient,
+		sinks:           sinks,
+	}
+}
+
+func (m *Manager) Init(ctx context.Context) error {
+	return m.do(ctx)
+}
+
+// Continuously updates the environment information.
+func (m *Manager) Start(ctx context.Context) error {
+	t := time.NewTicker(environmentProbeInterval)
+	defer t.Stop()
+	for range t.C {
+		if err := m.do(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *Manager) do(ctx context.Context) error {
+	env, err := m.probe(ctx)
+	if err != nil {
+		return err
+	}
+	for _, s := range m.sinks {
+		s.InjectEnvironment(env)
```

Comment:
Using the `Sink` implementation mentioned above would help, but as written these calls both block reconciliation and are blocked by reconciliation. With one reconciler it's  not a real concern, but it might make sense to use go routines and a waitgroup to reduce contention if more reconcilers use this.